mlir-sycl/doc/design/SYCLKernelSubmissionRepresentation.md

```diff
@@ -0,0 +1,510 @@
+overload of `sycl::handler::parallel_for`.
+-  We rely on
+`sycl::handler::StoreLambda` being inlined into `sycl::handler::parallel_for`,
+and the latter being inlined into the CGF.
+- The auto-generated *rounded range kernel* should be deactivated with `-Xclang
+  -fsycl-disable-range-rounding`.
+
+## The `sycl.host.schedule_kernel` operation
+
+The [`schedule_kernel`](#syclhostschedule_kernel-syclsyclhostschedulekernel)
+operation ties together the CGF's handler object, a symbol reference to a lambda
+function (i.e. the kernel), range information as passed to the `parallel_for`
+invocation, and the arguments captured by the lambda object.
+
+```mlir
+sycl.host.schedule_kernel %handler ->
+  @kernels::@k0[range %1](%2: !sycl_accessor_1_f32_rw_gb, %3)
+    : (!llvm.ptr, !llvm.ptr, !llvm.ptr, i32) -> ()
+```
+
+The operation allows to represent all `parallel_for` variants, i.e. executing
+over a range, a range with offset, and an nd-range. The arguments have the same
+granularity and types as described in the `kernel_signatures` data structure in
+the integration header. In particular, this means that accessors are "passed" as
+a pointer to an instance of one of the `sycl::accessor` classes, rather than
+being broken down into their parts. Accessor- and local accessor-arguments are
+annotated with their respective SYCL dialect type to preserve high-level
+information such as the access mode. Other *standard layout* objects such as
+generic structs and scalars are captured after reverting any copy optimizations
+done by the frontend (see [below](#captured-values)).
+
+At this point, the handler operand is merely an implementation detail which
+prevents the operation from being considered trivially dead without interfering
+with the dataflow analyses defined in context of the SYCL dialect. Also note
+that we do not model events or dependencies between different kernel launches
+yet — a future extension of the operation may have a result and operands
+representing events.
+
+## Raising approach
+
+Lots of high-level structure and information is lost during the lowering of the
+object-oriented SYCL source code to LLVM-IR. In order to recover the information
+required for the `schedule_kernel` operation, we employ a multi-stage pattern
+matching approach that is aided by source code annotations in the SYCL headers.
+The following subsections describe the intermediate steps in detail. The
+implementation resides in
+`polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp`. Note that
+example MLIR snippets have been manually edited to use descriptive names for
+values and other identifiers.
+
+### Command-group handler
+
+All raising patterns operate on the CGF itself or annotation known to reside in
+the CGF after mandatory inlining (see
+[above](#prerequisites--known-limitations)). Hence, the handler is first
+argument (after the `this` pointer) of the CGF.
+
+### Host constructors
```

Comment:
One thing to note for the host constructor raising is that it differs from other raisings in that changes the host code, removing the original call to the constructor and exception handling.